test/common/dashboards_data.go

```diff
@@ -0,0 +1,352 @@
+
+type dashboardDefinition struct {
+	Panels []panelDefinition `json:"panels"`
+}
+
+type dashboardDetailResponse struct {
+	Dashboard dashboardDefinition `json:"dashboard"`
+}
+
+type prometheusQueryResult struct {
+	Metric map[string]interface{} `json:"metric"`
+	Value  []interface{}          `json:"value"`
+}
+
+type prometheusQueryData struct {
+	Result []prometheusQueryResult `json:"result"`
+}
+
+type prometheusQueryResponse struct {
+	Status string              `json:"status"`
+	Data   prometheusQueryData `json:"data"`
+}
+
+var (
+	expectedServices = [...]string{
+		"3scale-admin-ui",
+		"3scale-developer-console-ui",
+		"apicurito-ui",
+		"codeready-ui",
+		"amq-service-broker",
+		"rhsso-ui",
+		"rhssouser-ui",
+		"3scale-system-admin-ui",
+		"webapp-ui",
+		"syndesis-ui",
+		"ups-ui",
+	}
+	dashboardsNames = [...]string{
+		"Endpoints Summary",
+		"Endpoints Detailed",
+		"Endpoints Report",
+		"Resource Usage By Pod",
+		"Resource Usage for Cluster",
+		"Resource Usage By Namespace",
+	}
+)
+
+// TestDashboardsData verifies that all dashboards are installed and all the graphs are filled with data
+func TestDashboardsData(t *testing.T, ctx *TestingContext) {
+	grafanaPodName, err := getMonitoringAppPodName("grafana", ctx)
+	if err != nil {
+		t.Fatal("failed to get grafana pod name", err)
+	}
+
+	prometheusPodName, err := getMonitoringAppPodName("prometheus", ctx)
+	if err != nil {
+		t.Fatal("failed to get prometheus pod name", err)
+	}
+
+	// retry the tests every minute for up to 10 minutes
+	monitoringTimeout := 10 * time.Minute
+	monitoringRetryInterval := 1 * time.Minute
+	err = wait.PollImmediate(monitoringRetryInterval, monitoringTimeout, func() (done bool, err error) {
+		expressions, err := getDashboardExpressions(grafanaPodName, prometheusPodName, ctx)
+		if err != nil {
+			t.Fatal("failed to get dashboard expressions", err)
+		}
+
+		queryOutputs, err := queryPrometheusMany(expressions, prometheusPodName, ctx)
+		if err != nil {
+			return false, err
+		}
+
+		var failedQueries []string
+
+		for i, queryOutput := range queryOutputs {
+			_, err := getPrometheusQueryResult(queryOutput)
+			if err != nil {
+				failedQueries = append(failedQueries, expressions[i])
+			}
+		}
+
+		for _, failedQuery := range failedQueries {
+			// not all containers define resource requests and limits -> expected failure
+			if strings.Contains(failedQuery, "kube_pod_container_resource_requests") ||
+				strings.Contains(failedQuery, "kube_pod_container_resource_limits") {
+				continue
+			}
+
+			// following query might fail, but dashboards are still visible -> ignoring
+			if strings.Contains(failedQuery, "probe_ssl_earliest_cert_expiry") {
+				continue
+			}
+
+			// completed pods don't use resources -> expected failure
+			matched, _ := regexp.Match(`pod=~'.*-(deploy|hook-pre|hook-post)'`, []byte(failedQuery))
+			if matched {
+				continue
+			}
+
+			t.Log("failed query:", failedQuery)
+			t.Log("waiting 1 minute before retrying")
```

Comment:
It would be nice to log all failed query before retrying

Reply:
Should be fixed now.